design-docs/improved-console.md

```diff
@@ -0,0 +1,422 @@
+
+This is opt-in initially by specifying the `org.gradle.console.parallel=true` Gradle property.
+
+### User Visible Changes
+The mainArea is displayed at the top for rendering important messages like warnings
+and errors.
+
+`TaskExecutionLogger` no longer sets logging headers. This prevents logging of task names
+in main area. To compensate for the user's loss of context, [WARN and ERROR logs can be 
+configured to give more context](#story-log-messages-can-be-configured-to-give-more-context),
+possibly setting the log category (and allowing `LogEvent` to accept a category)
+to the build operation description then configuring `StyledTextOutputBackedRenderer`
+to render category by default.
+
+In the case where a build process is not attached to a terminal, `OutputEvent`s are
+handled the same as previous versions of Gradle in that they are streamed to
+stdout and stderr with no filtering or throttling.
+ 
+The `ProgressLabel` is intended to give the user a glance-able indicator that tells
+them whether the build will be finished soon or not. 
+See ["Display build progress as a progress bar through colorized output"](#story-display-build-progress-as-a-progress-bar-through-colorized-output) below.
+
+The multiple progressArea lines are rendered results of a fixed list of `AbstractWorker`s. 
+It displays work-in-progress for each `AbstractWorker`. `BuildOperationDetails` is built
+with an optional worker ID to provide information about "where" an operation is running
+so that a mapping of worker to index of operation status line can be maintained.
+
+A `OperationStatusBarFormatter` can be used to customize the intra-line display of each 
+`ProgressOperation` chain (each `ProgressOperation` has a reference to its parent)
+
+### Implementation
+A `ParallelWorkReportingConsole` extends `Console` with the following operations: 
+ - `TextArea getMainArea()` (inherited from `Console`)
+ - `ProgressLabel getStatusBar()` (inherited from `Console`)
+ - `MultiStatusTextArea getProgressArea()`
+ 
+An `StatusBarFormatter` has operations:
+ - `String format(ProgressOperation operation)`
+
+`MultiStatusTextArea` extends `TextArea` with the following operations:
+ - `void update(ProgressOperations operations)`
+ - `void setStatusBarFormatter(StatusBarFormatter formatter)` 
+
+> NOTE: Plugins can provide progress indication by publishing `BuildOperations` (separate effort)
+
+Implementation is similar to `ConsoleBackedProgressRenderer` with throttling and a data
+structure for storing recent updates.
+
+Warnings and errors are displayed in the main area with the same format and styling as today.
+
+`FixedSizeMultiStatusTextArea` would be an implementation of `MultiStatusTextArea` that
+initializes a `List<LinkedList<ProgressOperation>>` of fixed size (defaults to `org.gradle.workers.max`
+ under `--parallel` or 1 otherwise). It would keep track of `ProgressOperation` chains 
+in-progress to prevent having more in-progress operations than max workers displayed.
```

Comment:
Sorry for the confusion. The intent here is that the List size is fixed to the size of the terminal but has no effect whatsoever on the actual work — just how many things are displayed.